### ISSUE TITLE
All numbers are imported as floats by XLS(X) importer

### ISSUE THREAD

**Issue by frafra** [OpenRefine/OpenRefine#5389]
### To Reproduce
1. Export a table containing an integer column

### Current Results
A cell showing the integer `123` is exported as `123.0`. This makes SQL exports fail with the following error: `MESSAGE:	123.0 is not compatible with column type :INT`.

### Expected Behavior
What you see is what you get: the software should not export cells in a different format.

### Versions
 - Operating System: Linux
 - Browser Version: Firefox
 - JRE or JDK Version: OpenJDK Runtime Environment 17.0.4+8-nixos
 - OpenRefine: 3.6.2

**Comment by elebitzero** [comment]
Hi @frafra,  I tried to reproduce this with the food-small.csv test data, and was not able to.  I exported two of the number columns Magnesium and Phosphorus as type INT in SQL Export, and it looks fine. No decimal point was added.

Example output:
```
CREATE TABLE food_small_csv (
Magnesium INT NULL,
Phosphorus INT NULL
);
INSERT INTO food_small_csv (Magnesium,Phosphorus) VALUES
( 2,24 ),
( 2,23 ),
( 0,3 ),
( 23,387 ),
( 24,451 ),
( 20,188 ),
( 20,347 ),
( 22,490 ),
( 28,512 ),
( 21,464 )
```

**Comment by frafra** [comment]
Could it be that is just Excel-files related?

**Comment by ostephens** [comment]
I don't think the issue is that an integer is being cast to a float, but rather that what looks like an integer in the OpenRefine UI is actually a float.

@elebitzero if you run a transformation like 
`value/1.0`

on the values in one of the columns before trying the export I think you'll see the issue manifest

**Comment by tfmorris** [comment]
This would be a lot easier to triage with example data which demonstrates the problem, so that folks don't have to guess.

**Comment by ostephens** [comment]
Try this project;
[sql-numbers-export.openrefine.tar.zip](https://github.com/OpenRefine/OpenRefine/files/10043411/sql-numbers-export.openrefine.tar.zip)

See in the history I've divided the value in row 2 by 1.0

Facet by star and select 'true' to isolate the second row. Try the SQL export with Numbers column set to export as INT. See error
Remove star facet and Facet by flag and select 'true' to isolate the first row. Try the SQL export with Numbers column set to export as INT. See success

**Comment by tfmorris** [comment]
Yes, I'm very familiar with how OpenRefine's number processing works . I have no idea of it's relevant to what @frafra reported, because they haven't included an example. I would fully expect the SQL exporter to export floats with decimal points in them. The simplest explanation is that the data isn't actually integer data, but an example would make this obvious.

**Comment by ostephens** [comment]
What I was trying to do is recreate the situation described by @frafra . I understand you are saying it would be good to get this directly from @frafra and anything else is speculation - fair enough. I was just trying to respond to your request to see some example data which demonstrates the problem - which I've done.

**Comment by frafra** [comment]
Here is the smallest example I could produce. I just opened LibreOffice and generated an XLSX file containing `integer` as first column and `123` under it.

[openrefine-issue-5389.xlsx](https://github.com/OpenRefine/OpenRefine/files/10068381/openrefine-issue-5389.xlsx)

I have the same issues with XLSX files generated by Microsoft Excel.

**Comment by ostephens** [comment]
@frafra thank you. Importing this seems to interpret the incoming number as a double rather than an integer. To solve your immediate issue you need to covert to an integer - e.g. you could first run `value.round()` on your column. As `round()` gives an integer output then you'll be able to export with the SQL Exporter to an INT column

I'm not sure what the correct fix for the underlying issue is here at the moment, but if you need an immediate working solution I think you'll find that `round()` works

**Comment by tfmorris** [comment]
Excel really only has two basic datatypes, strings and floating point numbers. Everything else, including dates, int vs float, percentages, scientific notation, etc is done with formatting strings. Numeric cells are always imported as floating point numbers ie Java `double`.

We could attempt to be more clever about trying to interpret format strings to decide if something should be imported as an integer, but there are some complicating factors like the "General" format (which is probably what's used here if the defaults were used).

**Comment by frafra** [comment]
Thanks for clarifying that. I guess #5397 could also be a valid workaround.

It would be nice to have a slightly smarter mechanism which evaluates casting to integer before casting to float. I am not familiar with the internals of OpenRefine, but I would say that a good approach is to see if a round-trip returns the same value. If that is the case for all the cells in the column, then the conversion is good, otherwise try another one, and use plain text as fallback.

**Comment by thadguidry** [comment]
@tfmorris not true anymore. Excel has very complex types (CT) now.  Here's some examples: https://interoperability.blob.core.windows.net/files/MS-XLSX/%5bMS-XLSX%5d.pdf

@tfmorris is correct in that the library we use, Apache POI, does parse your document correctly and assigned a Java type `double` to the cell value `123` [ref](https://github.com/apache/poi/blob/3890b7013b0e9c530b8d17fb95c86071cf7e4156/poi/src/main/java/org/apache/poi/ss/usermodel/CellValue.java#L47) since its underlying representation is that of a `real number` signified by the `t="n"` attribute in the Open Office XML format.  If it were represented as an `integer` then the attribute would be `t='i'`.

Notice your row's value `123` and its type `t="n"` in the underlying worksheet XML that I extracted:
```xml
    <sheetData>
        <row r="1" customFormat="false" ht="12.8" hidden="false" customHeight="false" outlineLevel="0" collapsed="false">
            <c r="A1" s="0" t="s">
                <v>0</v>
            </c>
        </row>
        <row r="2" customFormat="false" ht="12.8" hidden="false" customHeight="false" outlineLevel="0" collapsed="false">
            <c r="A2" s="0" t="n">
                <v>123</v>
            </c>
        </row>
    </sheetData>
```

**Comment by tfmorris** [comment]
I've combined the fixes for this and #5390 since they're basically the two different facets of the same problem (Excel format string handling). I'll leave this open until #5468 is merged.
